File: backend/app/models/banking.py
"""Connexions bancaries, comptes, saldos i traca de sincronitzacions."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import TYPE_CHECKING, Any

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, enum_column, money_column
from app.models.enums import ConnectionStatus, SyncStatus, SyncTrigger

if TYPE_CHECKING:
    from app.models.ledger import Ledger
    from app.models.transaction import Transaction


class BankConnection(Base, TimestampMixin):
    """Una autoritzacio d'Enable Banking: pot portar diversos comptes."""

    __tablename__ = "bank_connections"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(120), nullable=False, default="")
    aspsp_name: Mapped[str] = mapped_column(String(120), nullable=False)
    aspsp_country: Mapped[str] = mapped_column(String(2), nullable=False)
    psu_type: Mapped[str] = mapped_column(String(20), nullable=False, default="personal")
    eb_session_id: Mapped[str | None] = mapped_column(String(128), unique=True)
    eb_auth_state: Mapped[str | None] = mapped_column(String(128), index=True)
    status: Mapped[ConnectionStatus] = enum_column(
        ConnectionStatus, nullable=False, default=ConnectionStatus.PENDING
    )
    valid_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    last_sync_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    last_error: Mapped[str] = mapped_column(Text, nullable=False, default="")
    created_by_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))

    accounts: Mapped[list[Account]] = relationship(
        back_populates="connection", cascade="all, delete-orphan", lazy="selectin"
    )
    sync_runs: Mapped[list[SyncRun]] = relationship(
        back_populates="connection", cascade="all, delete-orphan"
    )

    @property
    def days_until_expiry(self) -> int | None:
        if self.valid_until is None:
            return None
        delta = self.valid_until - datetime.now(self.valid_until.tzinfo)
        return delta.days


class Account(Base, TimestampMixin):
    """Compte bancari, assignat a un llibre per l'usuari."""

    __tablename__ = "accounts"

    id: Mapped[int] = mapped_column(primary_key=True)
    connection_id: Mapped[int] = mapped_column(
        ForeignKey("bank_connections.id", ondelete="CASCADE"), nullable=False, index=True
    )
    ledger_id: Mapped[int | None] = mapped_column(
        ForeignKey("ledgers.id", ondelete="SET NULL"), index=True
    )
    eb_account_uid: Mapped[str] = mapped_column(String(128), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(160), nullable=False, default="")
    product: Mapped[str] = mapped_column(String(120), nullable=False, default="")
    iban: Mapped[str] = mapped_column(String(34), nullable=False, default="")
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="EUR")
    cash_account_type: Mapped[str] = mapped_column(String(20), nullable=False, default="")
    usage: Mapped[str] = mapped_column(String(20), nullable=False, default="")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    # Fins on hem arribat enrere amb l'historic i fins on hem sincronitzat endavant.
    history_start_date: Mapped[date | None] = mapped_column(Date)
    last_booked_date: Mapped[date | None] = mapped_column(Date)
    raw: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)

    connection: Mapped[BankConnection] = relationship(back_populates="accounts")
    ledger: Mapped[Ledger | None] = relationship(back_populates="accounts", lazy="selectin")
    balances: Mapped[list[Balance]] = relationship(
        back_populates="account", cascade="all, delete-orphan"
    )
    transactions: Mapped[list[Transaction]] = relationship(
        back_populates="account", cascade="all, delete-orphan"
    )

    @property
    def iban_masked(self) -> str:
        if len(self.iban) <= 8:
            return self.iban
        return f"{self.iban[:4]}…{self.iban[-4:]}"

    @property
    def display_name(self) -> str:
        return self.name or self.product or self.iban_masked


class Balance(Base):
    """Instantania de saldo, conservada per dibuixar l'evolucio."""

    __tablename__ = "balances"
    __table_args__ = (
        UniqueConstraint(
            "account_id", "balance_type", "reference_date", name="uq_balance_account_type_date"
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int] = mapped_column(
        ForeignKey("accounts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    balance_type: Mapped[str] = mapped_column(String(40), nullable=False)
    amount: Mapped[Decimal] = money_column(nullable=False)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="EUR")
    reference_date: Mapped[date] = mapped_column(Date, nullable=False)
    fetched_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    account: Mapped[Account] = relationship(back_populates="balances")


class SyncRun(Base):
    """Traca de cada sincronitzacio, tambe per controlar els limits de crides."""

    __tablename__ = "sync_runs"

    id: Mapped[int] = mapped_column(primary_key=True)
    connection_id: Mapped[int] = mapped_column(
        ForeignKey("bank_connections.id", ondelete="CASCADE"), nullable=False, index=True
    )
    trigger: Mapped[SyncTrigger] = enum_column(SyncTrigger, nullable=False)
    status: Mapped[SyncStatus] = enum_column(SyncStatus, nullable=False)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    accounts_synced: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    transactions_inserted: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    transactions_updated: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    error: Mapped[str] = mapped_column(Text, nullable=False, default="")

    connection: Mapped[BankConnection] = relationship(back_populates="sync_runs")

File: src/db/schema/banking.ts
/**
 * Bank connections, accounts, balances and sync attempts.
 *
 * Connections are **shared** between workspaces: the installation's
 * administrator manages them. Accounts, by contrast, are assigned to a
 * workspace (`ledgerId`), and until they have one their transactions are not
 * visible anywhere.
 */

import {
  boolean,
  date,
  foreignKey,
  index,
  integer,
  jsonb,
  pgTable,
  primaryKey,
  serial,
  text,
  unique,
  varchar,
} from "drizzle-orm/pg-core";

import { domainEnum, money, timestamps, tz } from "./columns.ts";
import type { ConnectionStatus, SyncStatus, SyncTrigger } from "./enums.ts";
import { ledgers } from "./ledgers.ts";
import { users } from "./users.ts";

export const bankConnections = pgTable(
  "bank_connections",
  {
    id: serial().notNull(),
    name: varchar({ length: 120 }).notNull(),
    aspspName: varchar("aspsp_name", { length: 120 }).notNull(),
    aspspCountry: varchar("aspsp_country", { length: 2 }).notNull(),
    psuType: varchar("psu_type", { length: 20 }).notNull(),
    /** The Enable Banking session opened after strong authentication. */
    ebSessionId: varchar("eb_session_id", { length: 128 }),
    /**
     * A single-use state tying the bank's callback to the connection that
     * started it. It is the secret protecting the callback route, which cannot
     * be authenticated because whoever arrives there comes from the bank.
     */
    ebAuthState: varchar("eb_auth_state", { length: 128 }),
    status: domainEnum<ConnectionStatus>().notNull(),
    /** Consent expiry: under PSD2, 90 days at most. */
    validUntil: tz("valid_until"),
    lastSyncAt: tz("last_sync_at"),
    lastError: text("last_error").notNull(),
    createdById: integer("created_by_id"),
    ...timestamps,
  },
  (t) => [
    primaryKey({ name: "pk_bank_connections", columns: [t.id] }),
    index("ix_bank_connections_eb_auth_state").on(t.ebAuthState),
    foreignKey({
      name: "fk_bank_connections_created_by_id_users",
      columns: [t.createdById],
      foreignColumns: [users.id],
    }).onDelete("set null"),
    unique("uq_bank_connections_eb_session_id").on(t.ebSessionId),
  ],
);

export const accounts = pgTable(
  "accounts",
  {
    id: serial().notNull(),
    connectionId: integer("connection_id").notNull(),
    /** Null = an account with no workspace yet: not visible from any of them. */
    ledgerId: integer("ledger_id"),
    ebAccountUid: varchar("eb_account_uid", { length: 128 }).notNull(),
    name: varchar({ length: 160 }).notNull(),
    product: varchar({ length: 120 }).notNull(),
    /** Personal data: only the masked form ever reaches a template. */
    iban: varchar({ length: 34 }).notNull(),
    currency: varchar({ length: 3 }).notNull(),
    cashAccountType: varchar("cash_account_type", { length: 20 }).notNull(),
    usage: varchar({ length: 20 }).notNull(),
    isActive: boolean("is_active").notNull(),
    historyStartDate: date("history_start_date"),
    /** How far transactions have been fetched; the sync window starts from here. */
    lastBookedDate: date("last_booked_date"),
    /**
     * The bank's whole response. It contains personal data (names, IBANs). It
     * is never rendered and never enters a fragment: see `AGENTS.md`.
     */
    raw: jsonb().notNull().$type<Record<string, unknown>>(),
    ...timestamps,
  },
  (t) => [
    primaryKey({ name: "pk_accounts", columns: [t.id] }),
    index("ix_accounts_connection_id").on(t.connectionId),
    index("ix_accounts_ledger_id").on(t.ledgerId),
    foreignKey({
      name: "fk_accounts_connection_id_bank_connections",
      columns: [t.connectionId],
      foreignColumns: [bankConnections.id],
    }).onDelete("cascade"),
    foreignKey({
      name: "fk_accounts_ledger_id_ledgers",
      columns: [t.ledgerId],
      foreignColumns: [ledgers.id],
    }).onDelete("set null"),
    unique("uq_accounts_eb_account_uid").on(t.ebAccountUid),
  ],
);

/** No `TimestampMixin`: whoever inserts sets `fetched_at`. */
export const balances = pgTable(
  "balances",
  {
    id: serial().notNull(),
    accountId: integer("account_id").notNull(),
    balanceType: varchar("balance_type", { length: 40 }).notNull(),
    amount: money().notNull(),
    currency: varchar({ length: 3 }).notNull(),
    referenceDate: date("reference_date").notNull(),
    fetchedAt: tz("fetched_at").notNull(),
  },
  (t) => [
    primaryKey({ name: "pk_balances", columns: [t.id] }),
    index("ix_balances_account_id").on(t.accountId),
    foreignKey({
      name: "fk_balances_account_id_accounts",
      columns: [t.accountId],
      foreignColumns: [accounts.id],
    }).onDelete("cascade"),
    unique("uq_balance_account_type_date").on(t.accountId, t.balanceType, t.referenceDate),
  ],
);

/** No `TimestampMixin`: each attempt carries its own `started_at`. */
export const syncRuns = pgTable(
  "sync_runs",
  {
    id: serial().notNull(),
    connectionId: integer("connection_id").notNull(),
    trigger: domainEnum<SyncTrigger>().notNull(),
    status: domainEnum<SyncStatus>().notNull(),
    startedAt: tz("started_at").notNull(),
    finishedAt: tz("finished_at"),
    accountsSynced: integer("accounts_synced").notNull(),
    transactionsInserted: integer("transactions_inserted").notNull(),
    transactionsUpdated: integer("transactions_updated").notNull(),
    error: text().notNull(),
  },
  (t) => [
    primaryKey({ name: "pk_sync_runs", columns: [t.id] }),
    index("ix_sync_runs_connection_id").on(t.connectionId),
    // Used to check the daily call limit the bank imposes.
    index("ix_sync_runs_started_at").on(t.startedAt),
    foreignKey({
      name: "fk_sync_runs_connection_id_bank_connections",
      columns: [t.connectionId],
      foreignColumns: [bankConnections.id],
    }).onDelete("cascade"),
  ],
);

export type BankConnection = typeof bankConnections.$inferSelect;
export type Account = typeof accounts.$inferSelect;
export type Balance = typeof balances.$inferSelect;
export type SyncRun = typeof syncRuns.$inferSelect;
export type NewSyncRun = typeof syncRuns.$inferInsert;
